contrib/kafka-humio-gateway/main.go

```diff
@@ -0,0 +1,288 @@
+	"syscall"
+	"time"
+
+	"github.com/go-yaml/yaml"
+	"github.com/Shopify/sarama"
+)
+
+var (
+	verbose  = false
+	configFile string
+	defaultConsumerGroup = "velociraptor-consumer"
+)
+
+
+type TransportConfig struct {
+	Kafka struct {
+		Brokers []string
+		Topics []string
+		ConsumerGroup string `yaml:"consumer_group"`
+	}
+	Humio struct {
+		EndpointUrl string `yaml:"endpoint_url"`
+		IngestToken string `yaml:"ingest_token"`
+	}
+}
+
+func init() {
+	flag.BoolVar(&verbose, "verbose", false, "Sarama logging")
+	flag.Parse()
+
+	args := flag.Args()
+
+	if len(args) != 1 {
+		log.Fatalf("error: config file missing")
+	}
+
+	configFile = args[0]
+}
+
+
+func main() {
+	/**
+	 * Setup a new Sarama consumer group
+	 */
+	consumer := Consumer{
+		ready: make(chan bool),
+	}
+
+	keepRunning := true
+
+	if verbose {
+		sarama.Logger = log.New(os.Stdout, "[sarama] ", log.LstdFlags)
+	}
+
+	body, err := os.ReadFile(configFile)
+
+	err = yaml.Unmarshal(body, &consumer.config)
+
+	if len(consumer.config.Kafka.Brokers) == 0 {
+		log.Fatalf("error: config has missing or empty kafka.brokers")
+	}
+
+	if len(consumer.config.Kafka.Topics) == 0 {
+		log.Fatalf("error: config has missing or empty kafka.topics")
+	}
+
+	if consumer.config.Kafka.ConsumerGroup == "" {
+		consumer.config.Kafka.ConsumerGroup = defaultConsumerGroup
+		log.Printf("warning: config missing `kafka.consumer_group'.  Using default `%s'", consumer.config.Kafka.ConsumerGroup)
+	}
+
+	log.Printf("Joining consumer group `%s'", consumer.config.Kafka.ConsumerGroup)
+
+	if consumer.config.Humio.EndpointUrl == "" {
+		log.Fatalf("error: config missing `humio.endpoint_url'")
+	}
+
+	_, err = url.ParseRequestURI(consumer.config.Humio.EndpointUrl)
+	if err != nil {
+		log.Fatalf("Humio Endpoint Url `%s' is not valid: %v", consumer.config.Humio.EndpointUrl, err)
+	}
+
+	log.Printf("Will post events to Humio Endpoint `%s'", consumer.config.Humio.EndpointUrl)
+
+	if consumer.config.Humio.IngestToken == "" {
+		log.Fatalf("error: config missing `humio.ingest_token'")
+	}
+
+
+	consumer.httpClient = http.Client{Timeout: time.Duration(1) * time.Second}
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	saramaConfig := sarama.NewConfig()
+	client, err := sarama.NewConsumerGroup(consumer.config.Kafka.Brokers, consumer.config.Kafka.ConsumerGroup, saramaConfig)
+	if err != nil {
+		log.Panicf("Error creating consumer group client: %v", err)
+	}
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go func() {
+		defer wg.Done()
+		for {
+			// `Consume` should be called inside an infinite loop, when a
+			// server-side rebalance happens, the consumer session will need to be
+			// recreated to get the new claims
+			if err := client.Consume(ctx, consumer.config.Kafka.Topics, &consumer); err != nil {
+				log.Panicf("Error from consumer: %v", err)
+			}
+			// check if context was cancelled, signaling that the consumer should stop
+			if ctx.Err() != nil {
+				return
+			}
+			consumer.ready = make(chan bool)
```

Comment:
This will be re-initializing the channel on every iteration, is this intended? Looking at the <-consumer.ready further down I'd expect to only want to initialize the consumer.ready chan just once?

Reply:
Actually consumer.ready is already initialized at the top of the function and I don't see where anything is being written to it. So perhaps you mean to do consumer.ready <- true here so unblock the main thread?

Reply:
Ok, this is actually correct since the channel is closed in the setup() callback of the consumer which would unblock the later read from the channel, thus ensuring the main thread continues only after the consumer has been initialized. However I find this rather non-obvious and probably a comment explaining this would be useful.

Reply:
TBH, I pulled this from a ConsumerGroup example. I agree it could be clearer.

Reply:
I'm going to replace this with a bool and a WaitGroup rather than this mess. The only reason the channel is re-made is because the setup callback expects it to be live. Nothing else uses it.